plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py

```diff
@@ -101,18 +101,23 @@ def destroy(self, dbg, sr):
     def attach(self, dbg, configuration):
         log.debug('{}: SR.attach: config={}'.format(dbg, configuration))
 
-        # ZFS automagically attaches a pool to a mountpoint on
-        # create/boot/etc, so we basically do nothing here but find
-        # this mountpoint
-
-        # FIXME: study how pools are mounted on boot and see if we
-        # could do real a/detach
+        pool_name = configuration["zpool"]
+        try:
+            mountpoint = zfsutils.pool_mountpoint(dbg, pool_name)
+        except Exception:
+            log.debug('{}: SR.attach: mountpoint for {} not found, import it'.format(dbg, pool_name))
+        else:
+            # SR.attach should be idempotent. So if the mountpoint is already
+            # mounted just return it.
+            return mountpoint
 
+        zfsutils.pool_import(dbg, pool_name)
         return zfsutils.pool_mountpoint(dbg, configuration["zpool"])
 
     def detach(self, dbg, sr):
-        # Nothing to unmount for now.
-        pass
+        meta = util.get_sr_metadata(dbg, 'file://' + sr)
+        # When exporting, all datasets within the pool are unmounted
```

Comment:
nitpicking: while this is true, we don't create other datasets inside the pool, only volumes, so only the pool itself gets unmounted.

Reply:
But volumes are dataset.

Reply:
As we disagree about what is a dataset ;) see; https://docs.oracle.com/cd/E18752_01/html/819-5461/gaypf.html
> A ZFS volume is a dataset that represents a block device

Reply:
OK, "zfs filesystems" and "volumes" are both datasets, note taken ;)

Nevertheless our "volume datasets" are not mounted, so the core of my remark still seems to hold :)